From 13102ae22584aeb119f13793a18e8c2591234239 Mon Sep 17 00:00:00 2001
From: Florian Westphal <fwestpha@redhat.com>
Date: Tue, 26 Aug 2014 23:47:41 -0400
Subject: [netdrv] veth: extend features to support tunneling

Message-id: <1409096861-26978-1-git-send-email-fwestpha@redhat.com>
Patchwork-id: 89256
O-Subject: [PATCH RHEL7.1 net] veth: extend features to support tunneling
Bugzilla: 1098138
RH-Acked-by: David S. Miller <davem@redhat.com>
RH-Acked-by: Jiri Pirko <jpirko@redhat.com>
RH-Acked-by: Daniel Borkmann <dborkman@redhat.com>
RH-Acked-by: Hannes Frederic Sowa <hannes@redhat.com>

BZ: 1098138
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7883963

Also avoids host tossing individual fragments
when running netperf UDP_STREAM with larger-than-mtu
sized packets (which resulted in extremely low tput).

commit 82d8189826d54740607e6a240e602850ef62a07d
Author: Eric Dumazet <edumazet@google.com>
Date:   Fri Oct 25 18:25:03 2013 -0700

    veth: extend features to support tunneling

    While investigating on a recent vxlan regression, I found veth
    was using a zero features set for vxlan tunnels.

    We have to segment GSO frames, copy the payload, and do the checksum.

    This patch brings a ~200% performance increase

    We probably have to add hw_enc_features support
    on other virtual devices.

    Signed-off-by: Eric Dumazet <edumazet@google.com>
    Cc: Alexei Starovoitov <ast@plumgrid.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Florian Westphal <fwestpha@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/veth.c b/drivers/net/veth.c
index cd38232..11c9ff1 100644
--- a/drivers/net/veth.c
+++ b/drivers/net/veth.c
@@ -256,6 +256,8 @@ static const struct net_device_ops veth_netdev_ops = {
 
 #define VETH_FEATURES (NETIF_F_SG | NETIF_F_FRAGLIST | NETIF_F_ALL_TSO |    \
          NETIF_F_HW_CSUM | NETIF_F_RXCSUM | NETIF_F_HIGHDMA | \
+         NETIF_F_GSO_GRE | NETIF_F_GSO_UDP_TUNNEL |     \
+         NETIF_F_GSO_IPIP | NETIF_F_GSO_SIT | NETIF_F_UFO |   \
          NETIF_F_HW_VLAN_CTAG_TX | NETIF_F_HW_VLAN_CTAG_RX | \
          NETIF_F_HW_VLAN_STAG_TX | NETIF_F_HW_VLAN_STAG_RX )
 
@@ -275,6 +277,7 @@ static void veth_setup(struct net_device *dev)
  dev->destructor = veth_dev_free;
 
  dev->hw_features = VETH_FEATURES;
+ dev->hw_enc_features = VETH_FEATURES;
 }
 
 /*
-- 
1.7.1